conn.go

```diff
@@ -791,60 +750,137 @@ func (c *Conn) isHandshakeCompletedSuccessfully() bool {
-			)
+		defer c.wg.Done()
+		err := c.fsm.Run(ctxHs, c, initialState)
+		if err != context.Canceled {
 			select {
-			case <-c.handshakeDoneSignal.Done():
-				return
-			case <-c.workerTicker.C:
-				isFinished, alertPtr, err = c.flightHandler(c)
-			case <-c.currFlight.workerTrigger:
-				isFinished, alertPtr, err = c.flightHandler(c)
+			case firstErr <- err:
+			default:
 			}
-
-			if alertPtr != nil {
-				if alertErr := c.notify(alertPtr.alertLevel, alertPtr.alertDescription); alertErr != nil {
-					err = fmt.Errorf("%v %v", err, alertErr)
+		}
+	}()
+	go func() {
+		defer c.wg.Done()
+		for {
+			if err := c.readAndBuffer(ctxRead); err != nil {
+				switch e := err.(type) {
+				case *errAlert:
+					if !e.IsFatalOrCloseNotify() {
+						if c.isHandshakeCompletedSuccessfully() {
+							// Pass the error to Read()
+							select {
+							case c.decrypted <- err:
+							case <-c.closed.Done():
+							}
+						}
+						err = nil // non-fatal alert must not stop read loop
+					}
+				case error:
+					switch err {
+					case context.DeadlineExceeded, context.Canceled, io.EOF:
+					default:
+						if c.isHandshakeCompletedSuccessfully() {
+							// Keep read loop and pass the read error to Read()
+							select {
+							case c.decrypted <- err:
+							case <-c.closed.Done():
+							}
+							err = nil
+						}
+					}
+				}
+				if err != nil {
+					select {
+					case firstErr <- err:
+					default:
+					}
+
+					if e, ok := err.(*errAlert); ok {
+						if e.IsFatalOrCloseNotify() {
+							_ = c.close(false)
+						}
+					}
+					return
 				}
-			}
-
-			switch {
-			case err != nil:
-				c.handshakeErr.store(err)
-				return
-			case c.readErr.load() != nil:
-				c.handshakeErr.store(c.readErr.load()) // Promote readErr to handshakeErr during handshake
-				return
-			case isFinished:
-				return // Handshake is complete
 			}
 		}
 	}()
-	c.currFlight.workerTrigger <- struct{}{}
+
+	select {
+	case err := <-firstErr:
+		cancelRead()
+		cancel()
+		return c.translateHandshakeCtxError(err)
+	case <-ctx.Done():
+		cancelRead()
+		cancel()
+		return c.translateHandshakeCtxError(ctx.Err())
+	case <-done:
+		return nil
+	}
 }
 
-func (c *Conn) close() error {
-	if c.connectionClosed.Err() == nil && c.handshakeErr.load() == nil {
-		_ = c.notify(alertLevelWarning, alertCloseNotify)
+func (c *Conn) translateHandshakeCtxError(err error) error {
+	switch err {
+	case context.Canceled:
+		if c.isHandshakeCompletedSuccessfully() {
+			return nil
+		}
+		return err
+	case context.DeadlineExceeded:
+		return errHandshakeTimeout
+	}
+	return err
+}
+
+func (c *Conn) close(byUser bool) error {
+	c.cancelHandshaker()
+	c.cancelHandshakeReader()
+	c.closed.Close()
+
+	if c.isHandshakeCompletedSuccessfully() && byUser {
+		_ = c.notify(context.Background(), alertLevelWarning, alertCloseNotify)
 	}
 
-	c.workerTicker.Stop()
-	c.handshakeDoneSignal.Close()
-	c.connectionClosed.Close()
-	return c.nextConn.Close()
+	c.closeLock.Lock()
```

Comment:
Getting/Setting things related to `close` should be in a function (or make everything atomic)

In the future developers will not know what is thread safe, could see races popping up accidentally.